DOC: extend the section on adding metadata to plans (#1327)
bluesky/bluesky: doc/source/metadata.rst

Capturing useful metadata is the main objective of bluesky. The more
information you can provide about what you are doing and why you are
doing it, the more useful bluesky and downstream data search and
analysis tools can be.

The term "metadata" can be a controversial term, one scientist's
"data" is another's "metadata" and classification is context
dependent.  The same exact information can be "data" in one
experiment, but "metadata" in a different experiment done on the exact
same hardware.
The `Document Model
<https://blueskyproject.io/event-model/data-model.html>`_ provides a framework
for deciding _where_ to record a particular piece of information.

There are somethings that we know a-priori before doing an experiment;
where are we? who is the user? what sample are we looking at? what did
the user just ask us to do?  These are all things that we can, in
principle, know independent of the control system.  These are the
prime candidates for inclusion in the `Start Document
<https://blueskyproject.io/event-model/data-model.html#run-start-document>`_.
Downstream DataBroker provides tools to do rich searches on this data.
The more information you can include the better.

> There is some information that we need to is nominally independent of

Reviewer: ```suggestion
There is some information that we need to that is nominally independent of
```

any particular device but we need to consult the controls system
about.  For example the location of important, but un-scanned motors

> or the configuration of filters.  If the values *should* be fixed over

Reviewer: ```suggestion
or the configuration of beam attenuators.  If the values *should* be fixed over
```

the course of the experiment then this it is a good candidate for
being a "baseline device" either via the `Supplemental pre-processor
<https://blueskyproject.io/bluesky/tutorial.html#baseline-readings-and-other-supplemental-data>`_
or explicitly in custom plans.  This will put the readings in a separate stream
(which is a peer to the "primary" data).  In principle, these values could be
read from the control system once and put into the Start document along with
the a-priori information, however that has several draw backs:

1. There is only ever 1 reading of the values so if they do drift during

>    data acquisition you will never know

Reviewer: ```suggestion
   data acquisition, you will never know.
```

2. We can not automatically capture information about the device like

>    we do for data in Events.  This include things like the datatype,

Reviewer: ```suggestion
   we do for data in Events.  This includes things like the datatype,
```

>    units, and shape of the value and any configuration information the hardware

Reviewer: ```suggestion
   units, and shape of the value and any configuration information about the hardware
```

   it is being read from.

A third class of information that can be called "metadata" is

> configuration information of a pieces of hardware.  This is things

Reviewer: ```suggestion
configuration information of pieces of hardware.  This is things
```

like the velocity of a motor or the integration time of a detector.

> These readings are embeded in the `Descriptor

Reviewer: ```suggestion
These readings are embedded in the `Descriptor
```

<https://blueskyproject.io/event-model/data-model.html#event-descriptor>`_
and are extracted from the hardware via the `read_configuration
<https://blueskyproject.io/bluesky/hardware.html#ReadableDevice.read_configuration>`_

> method the hardware.  We expect that these values will not change over

Reviewer: ```suggestion
method of the hardware.  We expect that these values will not change over
```

the course of the experiment so only read them once.

Information that does not fall into one of these categories, because
you expect it to change during the experiment, then we think that
should be treated as "data".  Either as an explicit part of the
experimental plan or via :ref:`async_monitoring`.

Adding to the Start Document
============================

When the RunEngine mints a Start document it includes structured data.  That

> information can be inject in via several mechanisms:

Reviewer: ```suggestion
information can be injected in via several mechanisms:
```

would override the individual 'purpose' metadata from the plan, marking all
three as purpose=test.

> If you define your own plans it is best practice have them take a keyword only

Reviewer: ```suggestion
If you define your own plans, it is best practice have them take a keyword only
```

or use any of the standard methods that apply to
`dictionaries in Python <https://docs.python.org/3/library/stdtypes.html#typesmapping>`_.

.. warning::

   In general we recommend against putting device readings in the Start
   document, however it is critical that you do not put device readings
   in this dictionary.  The value will remain until you change it and not
   track the state of the hardware.  This will result in recording incorrect
   data!

   This can be particularly dangerous if ``RE.md`` is backed by a

>    persistent data store (see next section)

Reviewer: ```suggestion
   persistent data store (see next section).
```
